test/integration/ctl/ctl_create_cr_test.go

```diff
@@ -0,0 +1,211 @@
+
+	// Create tmp directory and cd into it to store private key files
+	if err := os.Mkdir("tmp", 0777); err != nil {
+		t.Fatal(err)
+	}
+	if err := os.Chdir("tmp"); err != nil {
+		t.Fatal(err)
+	}
+
+	var (
+		cr1Name = "testcr-1"
+		cr2Name = "testcr-2"
+		cr3Name = "testcr-3"
+		cr4Name = "testcr-4"
+		ns1     = "testns-1"
+		ns2     = "testns-2"
+
+		testdataPath = "../testdata/"
+	)
+
+	// Create Namespaces
+	for _, ns := range []string{ns1, ns2} {
+		_, err := kubeClient.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: ns}}, metav1.CreateOptions{})
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+	tests := map[string]struct {
+		inputFile      string
+		inputArgs      []string
+		inputNamespace string
+		keyFilename    string
+
+		expErr         bool
+		expNamespace   string
+		expName        string
+		expKeyFilename string
+	}{
+		"v1alpha2 Certificate given": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr1Name},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expErr:         false,
+			expNamespace:   ns1,
+			expName:        cr1Name,
+			expKeyFilename: cr1Name + ".key",
+		},
+		"v1alpha3 Certificate given": {
+			inputFile:      testdataPath + "create_cr_v1alpha3_cert_with_ns1.yaml",
+			inputArgs:      []string{cr2Name},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expErr:         false,
+			expNamespace:   ns1,
+			expName:        cr2Name,
+			expKeyFilename: cr2Name + ".key",
+		},
+		"conflicting namespaces defined in flag and file": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr3Name},
+			inputNamespace: ns2,
+			keyFilename:    "",
+			expErr:         true,
+			expNamespace:   "",
+			expName:        "",
+			expKeyFilename: "",
+		},
+		"file passed in defines resource other than certificate": {
+			inputFile:      testdataPath + "create_cr_issuer.yaml",
+			inputArgs:      []string{cr4Name},
+			inputNamespace: ns1,
+			keyFilename:    "",
+			expErr:         true,
+			expNamespace:   "",
+			expName:        "",
+			expKeyFilename: "",
+		},
+		"path to file to store private key provided": {
+			inputFile:      testdataPath + "create_cr_cert_with_ns1.yaml",
+			inputArgs:      []string{cr1Name},
+			inputNamespace: ns1,
+			keyFilename:    "test.key",
+			expErr:         false,
+			expNamespace:   ns1,
+			expName:        cr1Name,
+			expKeyFilename: "test.key",
+		},
+	}
+
+	for name, test := range tests {
+		// Run ctl create cr command with input options
+		t.Run(name, func(t *testing.T) {
+			streams, _, _, _ := genericclioptions.NewTestIOStreams()
+
+			// Options to run create CR command
+			opts := &certificaterequest.Options{
+				CMClient:         cmCl,
+				RESTConfig:       config,
+				IOStreams:        streams,
+				CmdNamespace:     test.inputNamespace,
+				EnforceNamespace: test.inputNamespace != "",
```

Comment:
It’d be better if our tests didn’t have to encode so much ‘setup’ logic, as it means that this are of code in the actual ctl binary will not be tested. Is it possible for us to change the ‘inputs’ to be effectively the flag values, and use the Complete method on Options to set these? This will mean we are testing our actual ‘user facing API’ here then too (ie we will be effectively doing the same as an end user would do, and ensuring it continues to work as time goes on 🙂)

Reply:
That would have been my first choice as well, but I don't really understand how I would go about creating a cmdutil.Factory in the test, because Complete method takes in a Factory